src/desktop/space/mod.rs

```diff
@@ -0,0 +1,706 @@
+    pub fn unmap_window(&mut self, window: &Window) {
+        if let Some(map) = window.user_data().get::<WindowUserdata>() {
+            map.borrow_mut().remove(&self.id);
+        }
+        self.windows.shift_remove(window);
+    }
+
+    /// Iterate window in z-order back to front
+    pub fn windows(&self) -> impl DoubleEndedIterator<Item = &Window> {
+        self.windows.iter()
+    }
+
+    /// Get a reference to the window under a given point, if any
+    pub fn window_under<P: Into<Point<f64, Logical>>>(&self, point: P) -> Option<&Window> {
+        let point = point.into();
+        self.windows.iter().rev().find(|w| {
+            let bbox = window_rect(w, &self.id);
+            bbox.to_f64().contains(point)
+        })
+    }
+
+    /// Get a reference to the output under a given point, if any
+    pub fn output_under<P: Into<Point<f64, Logical>>>(&self, point: P) -> Option<&Output> {
+        let point = point.into();
+        self.outputs.iter().rev().find(|o| {
+            let bbox = self.output_geometry(o);
+            bbox.map(|bbox| bbox.to_f64().contains(point)).unwrap_or(false)
+        })
+    }
+
+    /// Returns the window matching a given surface, if any
+    pub fn window_for_surface(&self, surface: &WlSurface) -> Option<&Window> {
+        if !surface.as_ref().is_alive() {
+            return None;
+        }
+
+        self.windows
+            .iter()
+            .find(|w| w.toplevel().get_surface().map(|x| x == surface).unwrap_or(false))
+    }
+
+    /// Returns the layer matching a given surface, if any
+    pub fn layer_for_surface(&self, surface: &WlSurface) -> Option<LayerSurface> {
+        if !surface.as_ref().is_alive() {
+            return None;
+        }
+        self.outputs.iter().find_map(|o| {
+            let map = layer_map_for_output(o);
+            map.layer_for_surface(surface).cloned()
+        })
+    }
+
+    /// Returns the geometry of a [`Window`] including its relative position inside the Space.
+    pub fn window_geometry(&self, w: &Window) -> Option<Rectangle<i32, Logical>> {
+        if !self.windows.contains(w) {
+            return None;
+        }
+
+        Some(window_geo(w, &self.id))
+    }
+
+    /// Returns the bounding box of a [`Window`] including its relative position inside the Space.
+    pub fn window_bbox(&self, w: &Window) -> Option<Rectangle<i32, Logical>> {
+        if !self.windows.contains(w) {
+            return None;
+        }
+
+        Some(window_rect(w, &self.id))
+    }
+
+    /// Maps an [`Output`] inside the space.
+    ///
+    /// Can be safely called on an already mapped
+    /// [`Output`] to update its scale or location.
+    ///
+    /// The scale is the what is rendered for the given output
+    /// and may be fractional. It is independent from the integer scale
+    /// reported to clients by the output.
+    pub fn map_output<P: Into<Point<i32, Logical>>>(&mut self, output: &Output, scale: f64, location: P) {
```

Comment:
I'm not clear on how this interacts with damage tracking for that output. My intuition says that this should mark the whole output as damaged, but I'm not sure if that is what actually happens.